src/server/system_services/bucket_server.js

```diff
@@ -972,19 +972,65 @@ async function delete_bucket_lifecycle(req) {
  * LIST_BUCKETS
  *
  */
+
 async function list_buckets(req) {
-    const buckets_by_name = _.filter(
-        req.system.buckets_by_name,
-        async bucket => await req.has_s3_bucket_permission(bucket, "s3:ListBucket", req) && !bucket.deleting
-    );
-    return {
-        buckets: _.map(buckets_by_name, function(bucket) {
-            return {
-                name: bucket.name,
-                creation_date: bucket._id.getTimestamp().getTime()
-            };
-        })
-    };
+
+    const accessible_bucket_list = [];
+    const paged_bucket_list = [];
+    let next_index = 0;
+    let is_truncated = false;
+    const all_buckets = _.values(req.system.buckets_by_name);
+
+    let continuation_token = req.rpc_params?.continuation_token;
+    const max_buckets = req.rpc_params?.max_buckets;
+
+    dbg.error("max_buckets:", max_buckets, "continuation_token:", continuation_token);
+
+    for (const bucket of all_buckets) {
+        if (await req.has_s3_bucket_permission(bucket, "s3:ListBucket", req) && !bucket.deleting) {
+            accessible_bucket_list.push(bucket);
+        }
+    }
+    if (!max_buckets) {
+        const buckets = _.map(accessible_bucket_list, b => ({
+            name: b.name,
+            creation_date: b._id.getTimestamp().getTime()
+        }));
+        return {
+            buckets,
+        };
+    }
+    accessible_bucket_list.sort((a, b) => a._id.getTimestamp().getTime() - b._id.getTimestamp().getTime());
```

Comment:
If I remember correctly, the sort order should be lexicographic and not by creation date

Reply:
@dannyzaken We did talk about sorting  in lexicographical order but there are few cases which will not be covered.
1 - if there are few buckets like following 
a
aa
b
bb
c
cc
d
dd

Consider max length = 2
so after first call the continuation token would be "b" and max token would be 2. now in next call it will return "b" and "bb". Let say between these two call user created some buckets like "aaa" "aa1" "aaaa3". Then these buckets will never be listed in subsequent calls.

However, if we sort it with creation time then it will list all the above created buckets as the creation time for these buckets are latest. and we are sorting in increasing order of creation time.

I think this case is a valid test case to avoid using lexicographical order. As we are not saving the sorted order anywhere and not using it in any other usage, this would not break any other part.

Reply:
The problem you are trying to solve is not a requirement. 
I think we should be as consistent as possible with S3. The bucket list is always dynamic, and buckets are deleted and created constantly. You should also consider the application's needs—I believe that a lexicographic order is more what the UI expects to get from the server.

Reply:
done.